packages/shared/lib/services/connection.service.ts

```diff
@@ -881,18 +882,70 @@ class ConnectionService {
+                        if (success && connection) {
+                            const shouldRefresh = await this.shouldRefreshCredentials(
+                                connection,
+                                connection.credentials as OAuth2Credentials,
+                                providerConfig,
+                                template,
+                                false
+                            );
+                            if (!shouldRefresh) {
+                                return {
+                                    success: true,
+                                    error: null,
+                                    response: {
+                                        refreshed: true,
+                                        credentials: connection.credentials as
+                                            | OAuth2Credentials
+                                            | AppCredentials
+                                            | AppStoreCredentials
+                                            | OAuth2ClientCredentials
+                                    }
+                                };
+                            }
+                            freshConnection = connection;
+                        } else {
+                            throw error;
+                        }
                     }
-                    freshConnection = res.response;
+                } catch (err) {
+                    // lock acquisition might have timed out
+                    // but refresh might have been successfully performed by another execution
+                    // while we were waiting for the lock
+                    // so we check if the credentials were refreshed
+                    // if yes, we return the new credentials
+                    // if not, we actually fail the refresh
+                    const { success, response: connection } = await this.getConnection(connectionId, providerConfig.unique_key, environmentId);
+                    if (success && connection) {
+                        const shouldRefresh = await this.shouldRefreshCredentials(
+                            connection,
+                            connection.credentials as OAuth2Credentials,
+                            providerConfig,
+                            template,
+                            false
+                        );
+                        if (!shouldRefresh) {
+                            return {
+                                success: true,
+                                error: null,
+                                response: {
+                                    refreshed: true,
+                                    credentials: connection.credentials as OAuth2Credentials | AppCredentials | AppStoreCredentials | OAuth2ClientCredentials
+                                }
+                            };
+                        }
+                    }
+                    throw err;
```

Comment:
there is some code duplication but it is not trivial to consolidate the logci into one function. Let me know if you think I should spend more time trying to be more DRY

Reply:
I find this logic very confusing and hard to follow. This method has always been hard to follow IMO but if possible it would be great to make it more readable / easier to follow. Definitely push back if you think it would be too much of a time sink though

Reply:
The locking logic is indeed complicated. My quick attempt at trying to abstract the complexity was not very successful though, I was basically just moving the complexity around. I am happy to pair if you have some idea about how to improve the locking logic

Reply:
Made an attempt at removing the duplication [7ad0229](https://github.com/NangoHQ/nango/pull/2494/commits/7ad022989373db94678aa34db4de5b8b616c8a8e)
The concurrency/locking logic is still the same though